src/vm/internal/otable.cc

```diff
@@ -1,271 +1,131 @@
-  }
-#endif
-
-  ob->next_hash = obj_table[h];
-  obj_table[h] = ob;
-  objs_in_table++;
-
-  // for children()
-  base = _basename(ob->obname, &dummy);
-
-  ch = ObjHash(base);
-  ob->next_ch_hash = ch_table[ch];
-  ch_table[ch] = ob;
-  FREE_MSTR(base);
-  return;
+//return the children of key if any as a vector of object pointers(Values).
+ObjectTable::Vector ObjectTable::children(Key const & key) {
+    auto it = children_.find( basename(key) );
+    if( it != children_.end() ) {
+        return it->second;
+    }
+    else {
+        return Vector({});
+    }
 }
 
-/*
- * Remove an object from the table - generally called when it
- * is removed from the next_all list - i.e. in destruct.
- */
-
-void remove_object_hash(object_t *ob) {
-  int dummy;
-  object_t *s;
-  object_t *t = 0;
-  char *base;
-
-  s = find_obj_n(ob->obname); /* this sets h, and cycles the ob to the front */
-  if (!s) {
-    fatal("couldn't find object %s in obj_table", ob->obname);
-  }
-
-  DEBUG_CHECK1(s != ob, "Remove object \"/%s\": found a different object!", ob->obname);
-
-  obj_table[h] = ob->next_hash;
-  ob->next_hash = 0;
-  objs_in_table--;
-
-  base = _basename(ob->obname, &dummy);
-  ch = ObjHash(base);
-  FREE_MSTR(base);
-  s = ch_table[ch];
-  while (s && s != ob) {
-    t = s;
-    s = s->next_ch_hash;
-  }
-  if (!s) {
-    fatal("object not found in children list");
-  }
-  if (!t) {
-    ch_table[ch] = s->next_ch_hash;
-  } else {
-    t->next_ch_hash = s->next_ch_hash;
-  }
-  ob->next_ch_hash = 0;
-  return;
+//Attempt to find the object with name key in the object table and remove it
+//from both the objects map and the children map if found. Returns true if found,
+//otherwise false.
+//TODO: generalize this code by removing dependency on field obname.
+bool ObjectTable::remove(Key const & key) {
+    
+    auto it1 = objects_.find(key);
+    if( it1 == objects_.end() ) return false;
+    objects_.erase(it1);
+    //guaranteed to exist if object exists
+    //std::assert( children_.find( basename(key)) != children_.end() );
+    auto it2 = children_.find( basename(key) );
+    auto it3 = find_if( it2->second.begin(), it2->second.end(), [&key](Value v) -> bool { return v->obname == key; } );
+    //std::assert(it3 != it2->second.end())
+    it2->second.erase(it3);
+    if( it2->second.size() == 0 ) {
+      children_.erase(it2);
+    }
+    return true;
 }
 
-/*
- * Lookup an object in the hash table; if it isn't there, return null.
- * This is only different to find_object_n in that it collects different
- * stats; more finds are actually done than the user ever asks for.
- */
-
-static long user_obj_lookups = 0, user_obj_found = 0;
-
-object_t *lookup_object_hash(const char *s) {
-  object_t *ob = find_obj_n(s);
-
-  user_obj_lookups++;
-  if (ob) {
-    user_obj_found++;
-  }
-  return (ob);
+//write some information about memory usage and other statistics of the object table to out_buffer_t.
+//TODO: remove dependency on outbuffer_t here
```

Comment:
It's in the function definition, why do you need to remove dependency on outbuffer_t ?  again, you are putting some comment that is not relevant to current implementation, please go through the code and remove all those.